### PR TITLE
 Support register user defined CloudEventHandler when participant connect.

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#2122

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This change add an abstract class for CloudEventHandler and modified eventHandlerFactory to return instance of different implementation of abstract event handler.

### Tests

- [X] The following tests are written for this issue:

TestCloudEventCallbackProperty.java

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Comment by desaikomal** [comment]
can you please provide proper description of the change? thanks

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java [hunk 1]
```diff
@@ -832,7 +832,10 @@ public void connect() throws Exception {
       if (helixCloudProperty != null && helixCloudProperty.isCloudEventCallbackEnabled()) {
         _cloudEventListener =
             new HelixCloudEventListener(helixCloudProperty.getCloudEventCallbackProperty(), this);
-        CloudEventHandlerFactory.getInstance().registerCloudEventListener(_cloudEventListener);
+        LOG.info("Using handler: " + helixCloudProperty.getCloudEventHandlerClassName());
```
Comment (desaikomal) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java:835]: do we need the log at info level or debug?

File: helix-core/src/test/java/org/apache/helix/cloud/event/HelixTestCloudEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+public class HelixTestCloudEventHandler extends CloudEventHandler {
+  private static final int TIMEOUT = 900;  // second to timeout
+  private static ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+  @Override
+  public void registerCloudEventListener(CloudEventListener listener) {
+    super.registerCloudEventListener(listener);
+    System.out.println("in child register");
+  }
+
```
Comment (desaikomal) [inline helix-core/src/test/java/org/apache/helix/cloud/event/HelixTestCloudEventHandler.java:15]: nit: extra empty lines

File: helix-core/src/test/java/org/apache/helix/cloud/event/TestCloudEventCallbackProperty.java [hunk 3]
```diff
@@ -188,7 +193,7 @@ public void testUsingInvalidImplClassName() throws Exception {
     _helixManager.connect();
 
     // Manually trigger event
-    CloudEventHandlerFactory.getInstance()
+    ((CloudEventHandler) CloudEventHandlerFactory.getInstance(CloudEventHandler.class.getCanonicalName()))
```
Comment (desaikomal) [inline helix-core/src/test/java/org/apache/helix/cloud/event/TestCloudEventCallbackProperty.java:196]: can we add negative case where event handle is not registered and we call unregister etc

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java (hunk 1 above)
Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java:835]: I prefer to keep the info level as this log will only be write once when new Helix manager is created. Not logging anything will add a lot complexity in debug.

File: helix-core/src/test/java/org/apache/helix/cloud/event/TestCloudEventCallbackProperty.java (hunk 3 above)
Comment (xyuanlu) [reply helix-core/src/test/java/org/apache/helix/cloud/event/TestCloudEventCallbackProperty.java:196]: When event handle is not registered then unregister should be a no-op. It is valid to add a test case but it probably won't be a negative test case.

File: helix-core/src/test/java/org/apache/helix/cloud/event/MockCloudEventAwareHelixManager.java [hunk 4]
```diff
@@ -75,21 +75,32 @@ public MockCloudEventAwareHelixManager(HelixManagerProperty helixManagerProperty
...
   public void disconnect() {
     if (_cloudEventListener != null) {
-      CloudEventHandlerFactory.getInstance().unregisterCloudEventListener(_cloudEventListener);
-      _cloudEventListener = null;
+      try {
+        CloudEventHandlerFactory.getInstance(
+            _helixManagerProperty.getHelixCloudProperty().getCloudEventHandlerClassName())
+            .unregisterCloudEventListener(_cloudEventListener);
+      } catch (Exception e) {
+        System.out.println("Failed to unregister cloudEventListener.");
```
[the remark this answers is not in the document]
Comment (xyuanlu) [reply helix-core/src/test/java/org/apache/helix/cloud/event/MockCloudEventAwareHelixManager.java]: Updated.

File: helix-core/src/main/java/org/apache/helix/cloud/event/CloudEventHandlerFactory.java [hunk 5]
```diff
@@ -19,27 +19,35 @@
...
-  public static CloudEventHandler getInstance() {
-    if (INSTANCE == null) {
-      synchronized (CloudEventHandlerFactory.class) {
-        if (INSTANCE == null) {
-          INSTANCE = new CloudEventHandler();
-        }
+  public static AbstractEventHandler getInstance(String eventHandlerClassName)
+      throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+    synchronized (CloudEventHandlerFactory.class) {
+      if (INSTANCE_MAP.get(eventHandlerClassName) == null) {
```
[the remark this answers is not in the document]
Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/cloud/event/CloudEventHandlerFactory.java]: TFTR. Updated.

File: helix-core/src/test/java/org/apache/helix/cloud/event/HelixTestCloudEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+}
```
Comment (mgao0) [inline helix-core/src/test/java/org/apache/helix/cloud/event/HelixTestCloudEventHandler.java:16]: Could we test unregister as well? My imagination would be we disconnect the zk helix manager, and the listener should be automatically unregistered. It is missing in the current test, could you help add this test? Thanks!

File: helix-core/src/main/java/org/apache/helix/cloud/event/AbstractEventHandler.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package org.apache.helix.cloud.event;
+
```
[the remark this answers is not in the document]
Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/cloud/event/AbstractEventHandler.java:2]: TFTR. Updated.

File: helix-core/src/test/java/org/apache/helix/cloud/event/HelixTestCloudEventHandler.java (hunk 6 above)
Comment (xyuanlu) [reply helix-core/src/test/java/org/apache/helix/cloud/event/HelixTestCloudEventHandler.java:42]: Thanks for suggestion. Test added.

**Review by mgao0 — APPROVED** [review]
LGTM

**Comment by xyuanlu** [comment]
This change is ready to be merged. Approved by @mgao0 

Commit message:
**Support register user defined CloudEventHandler when participant connect.**
This change will load user defined CloudEventHandler dynamically when ZkHelixManager.connect(). User need to implement a EventController and pass the class name when initiating ZkHelixManager.
